File: internal/app/ports/mark.go
package ports

import "e-student/internal/app/domain"

type (
	MarkRepository interface {
		AddMark(mark *domain.Mark) error
		GetByStudent(studentID int) ([]domain.Mark, error)
		GetByClassGroup(classGroupID int) ([]domain.Mark, error)
	}

	MarkOutputSubject struct {
		ID   int    `json:"id"`
		Name string `json:"name"`
	}

	MarkOutputTeacher struct {
		ID      int    `json:"id"`
		Name    string `json:"name"`
		Surname string `json:"surname"`
	}

	MarkOutput struct {
		Comment string            `json:"comment,omitempty"`
		Value   float32           `json:"value"`
		Subject MarkOutputSubject `json:"subject"`
		Teacher MarkOutputTeacher `json:"teacher"`
	}

	SimpleMark struct {
		ID    int     `json:"id"`
		Value float32 `json:"value"`
	}

	MarkCreatePayload struct {
		SubjectID int     `json:"subjectID"`
		Value     float32 `json:"value"`
		StudentID int     `json:"studentID"`
		TeacherID int     `json:"teacherID"`
		Comment   string  `json:"comment"`
		Date      string  `json:"date"`
	}
	MarkService interface {
		CalculateAverage(marks []domain.Mark) float32
		GetByStudent(studentID int) ([]MarkOutput, error)
		GetByClassGroup(classGroupID int) ([]SimpleMark, error)
		CreateMark(payload MarkCreatePayload) error
	}
)

File: web/src/types/Mark.ts
export type AddMarkPayload = {
  subjectID: number
  value: number
  studentID: number
  teacherID: number
  comment?: string
  date?: string
}

export type StudentMark = {
  comment: string
  value: number
  subject: {
    id: number
    name: string
  }
  teacher: { id: number; name: string; surname: string }
}

type MarkMap = {
  1: string
  1.5: string
  2: string
  2.5: string
  3: string
  3.5: string
  4: string
  4.5: string
  5: string
  5.5: string
  6: string
}

export const markSimpleValue: MarkMap = {
  1: '1',
  1.5: '1+',
  2: '2',
  2.5: '2+',
  3: '3',
  3.5: '3+',
  4: '4',
  4.5: '4+',
  5: '5',
  5.5: '5+',
  6: '6'
}

export const markColor: MarkMap = {
  1: 'bg-danger',
  1.5: 'bg-danger',
  2: 'bg-warning',
  2.5: 'bg-warning',
  3: 'bg-warning',
  3.5: 'bg-primary',
  4: 'bg-primary',
  4.5: 'bg-pirmary',
  5: 'bg-primary',
  5.5: 'bg-primary',
  6: 'bg-success'
}
